lib/compress/zstd_opt.c

```diff
@@ -423,237 +424,241 @@ size_t ZSTD_compressBlock_opt_generic(ZSTD_CCtx* ctx,
+                } else {
+                    opt[cur].rep[2] = (opt[cur].off > 1) ? opt[cur-mlen].rep[1] : opt[cur-mlen].rep[2];
+                    opt[cur].rep[1] = (opt[cur].off > 0) ? opt[cur-mlen].rep[0] : opt[cur-mlen].rep[1];
+                    /* If opt[cur].off == ZSTD_REP_MOVE_OPT, then mlen != 1.
+                     * offset ZSTD_REP_MOVE_OPT is used for the special case
+                     * litLength == 0, where offset 0 means something special.
+                     * mlen == 1 means the previous byte was stored as a literal,
+                     * so they are mutually exclusive.
+                     */
+                    assert(!(opt[cur].off == ZSTD_REP_MOVE_OPT && mlen == 1));
+                    opt[cur].rep[0] = (opt[cur].off == ZSTD_REP_MOVE_OPT) ? (opt[cur-mlen].rep[0] - 1) : (opt[cur-mlen].rep[opt[cur].off]);
+            }   }
 
             best_mlen = minMatch;
-            {   U32 i, last_i = ZSTD_REP_CHECK + (mlen != 1);
-                for (i=(opt[cur].mlen != 1); i<last_i; i++) {  /* check rep */
-                    const S32 repCur = (i==ZSTD_REP_MOVE_OPT) ? (opt[cur].rep[0] - 1) : opt[cur].rep[i];
-                    if ( (repCur > 0) && (repCur < (S32)(inr-prefixStart))
-                       && (ZSTD_readMINMATCH(inr, minMatch) == ZSTD_readMINMATCH(inr - repCur, minMatch))) {
-                       mlen = (U32)ZSTD_count(inr+minMatch, inr+minMatch - repCur, iend) + minMatch;
-
-                       if (mlen > sufficient_len || cur + mlen >= ZSTD_OPT_NUM) {
-                            best_mlen = mlen; best_off = i; last_pos = cur + 1;
-                            goto _storeSequence;
-                       }
+            {   U32 const ll0 = (opt[cur].mlen != 1);
+                U32 const lastR = ZSTD_REP_CHECK + ll0;
+                U32 repCode4;  /* universal referential */
+                for (repCode4=ll0; repCode4<lastR; repCode4++) {  /* check rep */
+                    const S32 repCur = (repCode4==ZSTD_REP_MOVE_OPT) ? (opt[cur].rep[0] - 1) : opt[cur].rep[repCode4];
+                    if ( (repCur > 0) && (repCur < (S32)(inr-prefixStart))  /* within current mem segment */
+                      && (ZSTD_readMINMATCH(inr, minMatch) == ZSTD_readMINMATCH(inr - repCur, minMatch))) {
+                        U32 matchLength = (U32)ZSTD_count(inr+minMatch, inr+minMatch - repCur, iend) + minMatch;
+                        U32 const repCode3 = repCode4 - ll0;  /* contextual referential, depends on ll0 */
+                        assert(repCode3 < 3);
+
+                        if (matchLength > sufficient_len || cur + matchLength >= ZSTD_OPT_NUM) {
+                            best_mlen = matchLength;
+                            best_off = repCode4;
+                            last_pos = cur + 1;
+                            goto _shortestPath;
+                        }
 
-                       best_off = i - (opt[cur].mlen != 1);
```

Comment:
Why can you delete this `best_off` store? I see that you use `repCode3` below when getting/setting the price, but does anyone else depend on this store?

Reply:
I believe no one else reads it, but I just want to call it out.

Reply:
`best_off` was improperly used as a temporary variable to store the same value as `repCode3`.
It was not necessary, as the scope of the storage does not outlive this inner loop, hence it was replaced by a `repCode`, which has a narrower, and more accurate, scope.